update API for 2023

wongjustin99 opened the pull request:
updated API to work for current API

wongjustin99 committed:
update API for 2023

wongjustin99 committed:
updates for m1 macbook pro

kirjojenlukeminen commented:
Thank you very much for this PR.  This got me a lot closer to a working version.
I do still seem to run into an issue.
When I try `ruby main.rb` after editing `main.rb` to point to one of the books I've bought, I get the following error message:
```
/usr/local/lib/ruby/gems/3.2.0/gems/http-4.2.0/lib/http/response/body.rb:52:in `force_encoding': can't modify frozen String: "" (FrozenError)
	from /usr/local/lib/ruby/gems/3.2.0/gems/http-4.2.0/lib/http/response/body.rb:52:in `to_s'
	from [...]/books-dl-master/lib/books_dl/api.rb:93:in `login'
	from [...]/books-dl-master/lib/books_dl/api.rb:51:in `info'
	from [...]/books-dl-master/lib/books_dl/api.rb:26:in `initialize'
	from [...]/books-dl-master/lib/books_dl/downloader.rb:9:in `new'
	from [...]/books-dl-master/lib/books_dl/downloader.rb:9:in `initialize'
	from main.rb:11:in `new'
	from main.rb:11:in `<main>'
```
Do you happen to know how to solve this?

Thanks.

wongjustin99 commented:
It looks like your response body to the login is an empty string? You could run ruby's requests through a proxy like charles or proxyman to inspect what is happening.

kirjojenlukeminen commented:
> It looks like your response body to the login is an empty string? You could run ruby's requests through a proxy like charles or proxyman to inspect what is happening.

Thank you for your response.
I haven't quite figured out what's going on, but it seems I sometimes get a redirect response and http 4.2.0 doesn't seem to handle that very well:
```
# 4.2.0
irb(main):002:0> HTTP.get("https://db.books.com.tw/shopping/cart_list.php")
=> #<HTTP::Response/1.1 302 Found {"Date"=>"Mon, 28 Aug 2023 21:20:19 GMT", "Server"=>"Apache", "Set-Cookie"=>["bid=64ed0f9382845; expires=Wed, 27-Aug-2025 21:20:19 GMT; Max-Age=63072000; path=/; domain=books.com.tw; secure; HttpOnly", "ssid=64ed0f9382845.1693257619; expires=Mon, 28-Aug-2023 21:50:19 GMT; Max-Age=1800; path=/; domain=books.com.tw; secure; HttpOnly"], "Expires"=>"Mon, 26 Jul 1997 05:00:00 GMT", "Last-Modified"=>"Tue, 29 Aug 2023 05:20-1:19GMT", "Cache-Control"=>"max-age=0, private, no-store, no-cache, must-revalidate, post-check=0, pre-check=0", "Pragma"=>"no-cache", "Location"=>"https://www.books.com.tw/", "Content-Length"=>"0", "Content-Type"=>"text/html; charset=utf-8", "Connection"=>"close"}>                                        
irb(main):003:0> HTTP.follow.get("https://db.books.com.tw/shopping/cart_list.php")
/usr/local/lib/ruby/gems/3.2.0/gems/http-4.2.0/lib/http/response/body.rb:52:in `force_encoding': can't modify frozen String: "" (FrozenError)                                                            
        from /usr/local/lib/ruby/gems/3.2.0/gems/http-4.2.0/lib/http/response/body.rb:52:in `to_s'
        from /usr/local/lib/ruby/gems/3.2.0/gems/http-4.2.0/lib/http/response.rb:94:in `flush'
        from /usr/local/lib/ruby/gems/3.2.0/gems/http-4.2.0/lib/http/redirector.rb:59:in `perform'
        from /usr/local/lib/ruby/gems/3.2.0/gems/http-4.2.0/lib/http/client.rb:34:in `request'
        from /usr/local/lib/ruby/gems/3.2.0/gems/http-4.2.0/lib/http/chainable.rb:20:in `get'
        from (irb):3:in `<main>'                             
        from /usr/local/Cellar/ruby/3.2.2_1/lib/ruby/gems/3.2.0/gems/irb-1.6.2/exe/irb:11:in `<top (required)>'
        from /usr/local/opt/ruby/bin/irb:25:in `load'        
        from /usr/local/opt/ruby/bin/irb:25:in `<main>'  
# 5.1.0
irb(main):008:0> HTTP.get("https://db.books.com.tw/shopping/cart_list.php")
=> #<HTTP::Response/1.1 302 Found {"Date"=>"Mon, 28 Aug 2023 21:21:19 GMT", "Server"=>"Apache", "Set-Cookie"=>["bid=64ed0fcfc88b3; expires=Wed, 27-Aug-2025 21:21:19 GMT; Max-Age=63072000; path=/; domain=books.com.tw; secure; HttpOnly", "ssid=64ed0fcfc88b3.1693257679; expires=Mon, 28-Aug-2023 21:51:19 GMT; Max-Age=1800; path=/; domain=books.com.tw; secure; HttpOnly"], "Expires"=>"Mon, 26 Jul 1997 05:00:00 GMT", "Last-Modified"=>"Tue, 29 Aug 2023 05:21-1:19GMT", "Cache-Control"=>"max-age=0, private, no-store, no-cache, must-revalidate, post-check=0, pre-check=0", "Pragma"=>"no-cache", "Location"=>"https://www.books.com.tw/", "Content-Length"=>"0", "Content-Type"=>"text/html; charset=utf-8", "Connection"=>"close"}>
irb(main):009:0> HTTP.follow.get("https://db.books.com.tw/shopping/cart_list.php
")
=> #<HTTP::Response/1.1 200 OK {"Server"=>"nginx", "Date"=>"Mon, 28 Aug 2023 21:21:27 GMT", "Content-Type"=>"text/html; charset=UTF-8", "Transfer-Encoding"=>"chunked", "Connection"=>"close", "Vary"=>["Accept-Encoding", "Accept-Encoding,User-Agent"], "Cache-Control"=>"max-age=0", "Expires"=>"Mon, 28 Aug 2023 21:21:27 GMT"}>
```
Unfortunately, books-dl doesn't seem to work with http 5.1.0.
Any further suggestions?

Thanks.

wongjustin99 commented:
I used HTTP 4.2.0, but I think I might have actually removed the shopping cart URL check at one point when I saw my logins were going through (but not successfully being detected by this code). 

If you can see through the proxy that logins are working, try just removing the check (or fixing it to look at the response JSON).

kirjojenlukeminen commented:
> If you can see through the proxy that logins are working, try just removing the check (or fixing it to look at the response JSON).

Do you mean removing the line `return if logged?` from `def login`?
I'm not sure if it's related, but I'm apparently not seeing the redirects anymore.
However, the original problem persists.
I'm not using a proxy, but I just put some prints in the `get`method.
The last URL it tries to open is `https://appapi-ebook.books.com.tw/V1.6/CMSAPIApp/LoginURL?type=&device_id=&redirect_uri=https%3A%2F%2Fviewer-ebook.books.com.tw%2Fviewer%2Flogin.html`.  `response.status` is `200 OK`, but calling `to_s`on the body in `login_uri = JSON.parse(resp.body.to_s).fetch('login_uri')` still fails as before.

Do you have any further local changes that are not in this PR?

Thanks.

wongjustin99 commented:
No other local changes right now. 

During my experimentation I adjusted this method: 

```
    def logged?(headers={})
      @logged = begin
        response = get(CART_URL, headers)

        response.status == 200
      end
    end
```

to be the following: 

```
    def logged?(headers={})
      @logged = begin
        response = get(CART_URL, headers)

        return true
      end
    end
```

This was because I realised my login was succeeding (by observing the proxy traffic) and I didn't need to actually check, and I didn't feel like parsing the JSON response. 

Proxy traffic changes if you want to do to that: 


```
--- a/lib/books_dl/api.rb
+++ b/lib/books_dl/api.rb
@@ -136,7 +136,7 @@ module BooksDL

     def get(url, headers = {})
       headers = build_headers({ Cookie: cookie }, headers)
-      response = HTTP.headers(headers).get(url)
+      response = HTTP.via("localhost", 8888).headers(headers).get(url)

       if response.status >= 400
         file_name = URI(url).path.split('/').last
@@ -149,7 +149,7 @@ module BooksDL

     def post(url, data = {}, headers = {})
       headers = build_headers({ Cookie: cookie }, headers)
-      response = HTTP.headers(headers).post(url, data)
+      response = HTTP.via("localhost", 8888).headers(headers).post(url, data)
       save_cookie(response)
```

Tristan-Lo commented:
Thank you Justin, the 2023 api update works well for me.

JLBBHK commented:
I ran the script in windows and got the following error. Do you know why?

Resolving dependencies...
C:/Ruby33-x64/lib/ruby/3.3.0/bundler/definition.rb:570:in `materialize': Could not find http-4.2.0, nokogiri-1.10.5, rubyzip-2.0.0, awesome_print-1.8.0, byebug-11.0.1, rspec-3.9.0, addressable-2.7.0, http-cookie-1.0.3, http-form_data-2.1.1, http-parser-1.2.1, mini_portile2-2.4.0, rspec-core-3.9.0, rspec-expectations-3.9.0, rspec-mocks-3.9.0, public_suffix-4.0.1, domain_name-0.5.20190701, ffi-compiler-1.0.1, rspec-support-3.9.0, diff-lcs-1.3, unf-0.1.4, ffi-1.11.2, rake-13.0.1, unf_ext-0.0.7.6 in locally installed gems (Bundler::GemNotFound)
        from C:/Ruby33-x64/lib/ruby/3.3.0/bundler/definition.rb:272:in `specs_for'
        from C:/Ruby33-x64/lib/ruby/3.3.0/bundler/runtime.rb:18:in `setup'
        from C:/Ruby33-x64/lib/ruby/3.3.0/bundler.rb:164:in `setup'
        from C:/Ruby33-x64/lib/ruby/3.3.0/bundler.rb:187:in `require'
        from C:/Users/Jason/Tools/books-dl-master/lib/books_dl.rb:8:in `<top (required)>'
        from main.rb:1:in `require_relative'
        from main.rb:1:in `<main>'

Tristan-Lo commented:
@JLBBHK I was using ruby2.6.5 and bundler:2.0.2. Newer ruby does not work. 
also, have you updated the main.rb according to [e852f7a](https://github.com/lostfilm/books-dl/pull/3/commits/e852f7a873d55da78be9d115ceb3d0e997608e00)?

JLBBHK commented:
Thank you. Let me install 2.6.5. also double check the main.rb and try again.

JLBBHK commented:
I am sorry that I am not familiar with this... I download the latest codes, installed ruby 2.6.5 and bundler 2.0.2, but got the follow error. Can anyone please help?

Traceback (most recent call last):
        10: from main.rb:1:in `<main>'
         9: from main.rb:1:in `require_relative'
         8: from C:/Users/Jason/Tools/books-dl-master/lib/books_dl.rb:8:in `<top (required)>'
         7: from C:/Ruby/lib/ruby/gems/2.6.0/gems/bundler-2.0.2/lib/bundler.rb:114:in `require'
         6: from C:/Ruby/lib/ruby/gems/2.6.0/gems/bundler-2.0.2/lib/bundler.rb:109:in `setup'
         5: from C:/Ruby/lib/ruby/gems/2.6.0/gems/bundler-2.0.2/lib/bundler/runtime.rb:20:in `setup'
         4: from C:/Ruby/lib/ruby/gems/2.6.0/gems/bundler-2.0.2/lib/bundler/definition.rb:237:in `specs_for'
         3: from C:/Ruby/lib/ruby/gems/2.6.0/gems/bundler-2.0.2/lib/bundler/definition.rb:170:in `specs'
         2: from C:/Ruby/lib/ruby/gems/2.6.0/gems/bundler-2.0.2/lib/bundler/spec_set.rb:81:in `materialize'
         1: from C:/Ruby/lib/ruby/gems/2.6.0/gems/bundler-2.0.2/lib/bundler/spec_set.rb:81:in `map!'
C:/Ruby/lib/ruby/gems/2.6.0/gems/bundler-2.0.2/lib/bundler/spec_set.rb:87:in `block in materialize': Could not find rake-13.0.1 in any of the sources (Bundler::GemNotFound)

Tristan-Lo commented:
Have you installed the required gems by running bundle install?
________________________________
From: JLBBHK ***@***.***>
Sent: Tuesday, January 23, 2024 6:49:57 PM
To: lostfilm/books-dl ***@***.***>
Cc: Tristan Lo ***@***.***>; Comment ***@***.***>
Subject: Re: [lostfilm/books-dl] update API for 2023 (PR #3)


I am sorry that I am not familiar with this... I download the latest codes, installed ruby 2.6.5 and bundler 2.0.2, but got the follow error. Can anyone please help?

Traceback (most recent call last):
10: from main.rb:1:in <main>' 9: from main.rb:1:in require_relative'
8: from C:/Users/Jason/Tools/books-dl-master/lib/books_dl.rb:8:in <top (required)>' 7: from C:/Ruby/lib/ruby/gems/2.6.0/gems/bundler-2.0.2/lib/bundler.rb:114:in require'
6: from C:/Ruby/lib/ruby/gems/2.6.0/gems/bundler-2.0.2/lib/bundler.rb:109:in setup' 5: from C:/Ruby/lib/ruby/gems/2.6.0/gems/bundler-2.0.2/lib/bundler/runtime.rb:20:in setup'
4: from C:/Ruby/lib/ruby/gems/2.6.0/gems/bundler-2.0.2/lib/bundler/definition.rb:237:in specs_for' 3: from C:/Ruby/lib/ruby/gems/2.6.0/gems/bundler-2.0.2/lib/bundler/definition.rb:170:in specs'
2: from C:/Ruby/lib/ruby/gems/2.6.0/gems/bundler-2.0.2/lib/bundler/spec_set.rb:81:in materialize' 1: from C:/Ruby/lib/ruby/gems/2.6.0/gems/bundler-2.0.2/lib/bundler/spec_set.rb:81:in map!'
C:/Ruby/lib/ruby/gems/2.6.0/gems/bundler-2.0.2/lib/bundler/spec_set.rb:87:in `block in materialize': Could not find rake-13.0.1 in any of the sources (Bundler::GemNotFound)

—
Reply to this email directly, view it on GitHub<https://github.com/lostfilm/books-dl/pull/3#issuecomment-1905780726>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/A3WB44VM4D7JDRFJGLVI7R3YP6IVLAVCNFSM6AAAAAA3H45PGKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMBVG44DANZSGY>.
You are receiving this because you commented.Message ID: ***@***.***>

 ____________________________________________________________
Electronic mail messages entering and leaving Arup business systems are scanned for viruses and acceptability of content.

JLBBHK commented:
@Tristan-Lo I bundle install and there is a gem caused error while installing. Do you have idea?

Installing byebug 11.0.1 with native extensions
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    current directory: C:/Ruby/lib/ruby/gems/2.6.0/gems/byebug-11.0.1/ext/byebug
C:/Ruby/bin/ruby.exe -I C:/Ruby/lib/ruby/2.6.0 -r
./siteconf20240123-25720-ckswml.rb extconf.rb
creating Makefile

current directory: C:/Ruby/lib/ruby/gems/2.6.0/gems/byebug-11.0.1/ext/byebug
make "DESTDIR=" clean
current directory: C:/Ruby/lib/ruby/gems/2.6.0/gems/byebug-11.0.1/ext/byebug
make "DESTDIR="
make failedNo such file or directory - make "DESTDIR="

Gem files will remain installed in
C:/Ruby/lib/ruby/gems/2.6.0/gems/byebug-11.0.1 for inspection.
Results logged to
C:/Ruby/lib/ruby/gems/2.6.0/extensions/x86-mingw32/2.6.0/byebug-11.0.1/gem_make.out

An error occurred while installing byebug (11.0.1), and Bundler cannot continue.
Make sure that `gem install byebug -v '11.0.1' --source 'https://rubygems.org/'`
succeeds before bundling.

In Gemfile:
  byebug

JLBBHK commented:
Tried that with Mac, get stuck on installing byebug.

tristantylo commented:
sorry, run on PC, will try on mac and get back to you

JLBBHK commented:
> sorry, run on PC, will try on mac and get back to you

No worry bud. I finally got it working. Had a clean re-install of ruby and MSYS2. Installed byebug before bundle install works. Thanks for your help.

JLBBHK commented:
Well. However, I couldn't login successfully. Is this the correct sequence? 
1. login id (tried both id name and email)
2. password
3. It prompts open failure and asked me to check captcha.
4. I did and enter the captcha. But login failed. :(

Do I need to login first from browser? I logged out and close browser while doing this.

Tristan-Lo commented:
Correct, after **ruby main.rb**
user name 
password
captcha (case insensitive)
then it should start downloading.
check your api.rb with [e852f7a](https://github.com/lostfilm/books-dl/pull/3/commits/e852f7a873d55da78be9d115ceb3d0e997608e00)?
and in main.rb update the book_id to a book that you have.
also logout everywhere else before running since books.com.tw prevents multiple login.

JLBBHK commented:
@Tristan-Lo I updated api.rb as in e852f7a, now works fine. Thank you so much :)

irisjae commented:
With this fork and tpai's fork, I got gibberish on the encrypted contents.

For some reason (file structures changed?), I found that the paths of the book contents provided to `lib/books_dl/files/content.rb` had no parent directory (like "file.xhtml" as opposed to "dir/file.xhtml")

This caused the detected dirname to be '.' and the paths used by the downloader to be like http://something/./file.xhtml rather than http://something/file.xhtml. This downloads successfully, but will produce an incorrect decryption key, causing the gibberish

If you encounter this, https://github.com/irisjae/books-dl/commit/b747dce260a25415f5eaf24274e612592b0c04aa may be useful..

(On another note, given that the files provided are all merely XOR encoded, the keys can alternatively be easily found without reimplementing the key generation algorithm. If given a well-known file e.g. the contents, the correct file information should not be hard to reconstruct, especially since the format is well-formed XML. Another more generic method is to request some file three separate times, so if we let X be the file content, and A, B, C be respectively the three encoding keys, we have:
```
code_A = X xor A
code_B = X xor B
code_C = X xor C
```
, from which we have
```
code_A xor code_B = A xor B
code_B xor code_C = B xor C
code_C xor code_A = C xor A
```
, from which we can apply Gaussian elimination or any other linear solver to find all three keys. Of course, these methods assume that the chosen file is at least as long as the key.)

JLBBHK commented:
I recently got this error, the epub failed to be created, can anyone please help?

C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/rubyzip-2.0.0/lib/zip/output_stream.rb:36:in `initialize': Invalid argument @ rb_sysopen - ./E050194176_reflowable_normal_IP時代必備的創作指南──網路漫畫、網路小說作者最好奇的58個FAQ 스토리, 꼭 그래야  할까? 다르게 쓰고 싶은 웹툰-웹소설 작가를 위한 가이드.epub20240807-2076-jo93bw (Errno::EINVAL)

cytalya commented:
Hello, sorry it is my first time working on this, can someone pls help?
I installed ruby 3.3.5 finally, and don't know if my location is correct, i just drag the file there

also, do I have to change the book id in Api.rb ?? after updated with the 2023

ruby /Users/cyt/Documents/GitHub/books-dl/main.rb 
/usr/local/Cellar/ruby/3.3.5/lib/ruby/3.3.0/bundler/shared_helpers.rb:23:in `default_gemfile': Could not locate Gemfile (Bundler::GemfileNotFound)
	from /usr/local/Cellar/ruby/3.3.5/lib/ruby/3.3.0/bundler/shared_helpers.rb:28:in `default_lockfile'
	from /usr/local/Cellar/ruby/3.3.5/lib/ruby/3.3.0/bundler.rb:486:in `default_lockfile'
	from /usr/local/Cellar/ruby/3.3.5/lib/ruby/3.3.0/bundler.rb:230:in `definition'
	from /usr/local/Cellar/ruby/3.3.5/lib/ruby/3.3.0/bundler.rb:158:in `setup'
	from /usr/local/Cellar/ruby/3.3.5/lib/ruby/3.3.0/bundler.rb:212:in `require'
	from /Users/cyt/Documents/GitHub/books-dl/lib/books_dl.rb:8:in `<top (required)>'
	from /Users/cyt/Documents/GitHub/books-dl/main.rb:1:in `require_relative'
	from /Users/cyt/Documents/GitHub/books-dl/main.rb:1:in `<main>'

Tristan-Lo commented:
@cytalya Newer ruby does not work. Or let me know if it does now. Thx. 
> @JLBBHK I was using ruby2.6.5 and bundler:2.0.2. Newer ruby does not work. also, have you updated the main.rb according to [e852f7a](https://github.com/lostfilm/books-dl/pull/3/commits/e852f7a873d55da78be9d115ceb3d0e997608e00)?

viccat54 commented:
I used this version successfully, thank you! @wongjustin99

sueliu73 commented:
Yes, thanks to @wongjustin99, as your branch is 2 commits ahead of lostfilm/books-dl:master, therefore it works well, so thrilled, thank you all and @lostfilm

weaponmasterjax commented:
Now we get 
Traceback (most recent call last):
        6: from main.rb:12:in `<main>'
        5: from /home/jax/books-dl/lib/books_dl/downloader.rb:21:in `perform'
        4: from /home/jax/books-dl/lib/books_dl/downloader.rb:34:in `job'
        3: from /home/jax/books-dl/lib/books_dl/downloader.rb:21:in `block in perform'
        2: from /home/jax/books-dl/lib/books_dl/downloader.rb:39:in `fetch_container_file'
        1: from /home/jax/books-dl/lib/books_dl/api.rb:41:in `fetch'
/home/jax/books-dl/lib/books_dl/utils.rb:11:in `generate_key': undefined method `[]' for nil:NilClass (NoMethodError)
after passing in the cookies from a logged in  browser. So that, if I understand correctly, is due to the mechanism that prevents multiple logins. Since they updated the captcha, we can't login just with userpass. Maybe we need to rewrite this using selenium web driver so that the user can solve the captcha by themselves?
